### PR TITLE
feat: repository controller supports sync serviceaccount

### PR DESCRIPTION
**Pull request by 0xff-dev**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it
impl Rating.

#### Which issue(s) this PR fixes

Fix: https://github.com/kubebb/core/issues/250

### CODE REVIEW

File: api/v1alpha1/repository.go [hunk 1]
```diff
@@ -58,3 +68,63 @@ func GetImageOverridePath() []string {
 	}
 	return strings.Split(v, ":")
 }
+
+func RatingControllerEnable() bool {
```
Comment (bjwswang) [inline api/v1alpha1/repository.go:72]: `RatingControllerEnable` -> `RatingControllerEnabled`

File: api/v1alpha1/repository.go [hunk 2]
```diff
@@ -58,3 +68,63 @@ func GetImageOverridePath() []string {
...
+	return nil
+}
+
+func RemoveSubjectFromClusterRoleBinding(ctx context.Context, c client.Client, namespace string) error {
+	if !RatingControllerEnable() {
+		return nil
+	}
+
+	clusterRoleBinding := os.Getenv(PipelineRunClusterRoleBindingEnv)
+	serviceAccount := os.Getenv(PipelineRunServiceAccountEnv)
```
Comment (Abirdcfly) [inline api/v1alpha1/repository.go:110]: if clusterRoleBinding or serviceAccount is empty, just return nil here?

File: api/v1alpha1/repository.go [hunk 3]
```diff
@@ -58,3 +68,63 @@ func GetImageOverridePath() []string {
...
+func RatingControllerEnable() bool {
+	return os.Getenv(RatingEnableEnv) == "true"
+}
+
+func AddSubjectToClusterRoleBinding(ctx context.Context, c client.Client, namespace string) error {
+	if !RatingControllerEnable() {
+		return nil
+	}
+
+	clusterRoleBinding := os.Getenv(PipelineRunClusterRoleBindingEnv)
```
Comment (bjwswang) [inline api/v1alpha1/repository.go:81]: Should we define default names for clustrole/clusterrolebinding/service account

File: controllers/repository_controller.go [hunk 4]
```diff
@@ -205,6 +214,58 @@ func (r *RepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	}
+	return manager.Complete(r)
+}
+
+func (r RepositoryReconciler) CheckOrCreateServiceAccount(ctx context.Context, namespace string) error {
+	if !corev1alpha1.RatingControllerEnable() {
+		return nil
+	}
+
+	ratingUsedServiceAccount := os.Getenv(corev1alpha1.PipelineRunServiceAccountEnv)
```
Comment (Abirdcfly) [inline controllers/repository_controller.go:256]: again, if empty return?

**Review by Abirdcfly — APPROVED** [review]

File: api/v1alpha1/repository.go [hunk 5]
```diff
@@ -58,3 +68,63 @@ func GetImageOverridePath() []string {
...
+	if err := c.Get(ctx, types.NamespacedName{Name: clusterRoleBinding}, &crb); err != nil {
+		return err
+	}
+
+	index, length := 0, len(crb.Subjects)
+	for idx := 0; idx < length; idx++ {
+		if crb.Subjects[idx].Kind == "ServiceAccount" && crb.Subjects[idx].Name == serviceAccount && crb.Subjects[idx].Namespace == namespace {
+			continue
+		}
+		crb.Subjects[index] = crb.Subjects[idx]
```
Comment (bjwswang) [inline api/v1alpha1/repository.go:121]: Cool! @0xff-dev

File: controllers/repository_controller.go [hunk 6]
```diff
@@ -205,6 +214,58 @@ func (r *RepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+						}
+					}
+				}
+			},
+		})
+	}
+	return manager.Complete(r)
+}
+
+func (r RepositoryReconciler) CheckOrCreateServiceAccount(ctx context.Context, namespace string) error {
```
Comment (bjwswang) [inline controllers/repository_controller.go:251]: better to name it `EnsureRatingServiceAccount` with `Rating` specified

File: api/v1alpha1/repository.go [hunk 7]
```diff
@@ -58,3 +75,96 @@ func GetImageOverridePath() []string {
...
+
+func GetRatingServiceAccount() string {
+	if r := os.Getenv(RatingServiceAccountEnv); r != "" {
+		return r
+	}
+	return DefaultRatingServiaceAccount
+}
+
+func GetRatingClusterRole() string {
+	if r := os.Getenv(RatingClusterRoleEnv); r != "" {
```
Comment (Abirdcfly) [inline api/v1alpha1/repository.go:105]: same with line 98

File: api/v1alpha1/repository.go [hunk 8]
```diff
@@ -58,3 +75,96 @@ func GetImageOverridePath() []string {
...
+		return r
+	}
+	return DefaultRatingClusterRole
+}
+
+func GetRatingClusterRoleBinding() string {
+	if r := os.Getenv(RatingClusterRoleBindingEnv); r != "" {
```
Comment (Abirdcfly) [inline api/v1alpha1/repository.go:112]: same with line 98

File: api/v1alpha1/repository.go [hunk 9]
```diff
@@ -58,3 +75,96 @@ func GetImageOverridePath() []string {
...
+	}
+
+	clusterRolebingName := GetRatingClusterRoleBinding()
+	if _, err := c.RbacV1().ClusterRoleBindings().Get(context.Background(), clusterRolebingName, metav1.GetOptions{}); err != nil {
+		panic(err)
+	}
+}
+
+func RatingEnabled() bool {
+	return os.Getenv(RatingEnableEnv) == "true"
```
Comment (Abirdcfly) [inline api/v1alpha1/repository.go:94]: "k8s.io/utils/env".GetBool(RatingEnableEnv, false) ?

**Review by Abirdcfly — DISMISSED** [review]

**Review by Abirdcfly — APPROVED** [review]

**Review by bjwswang — APPROVED** [review]

**Pull request merged**